staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,247 @@
+	begin := time.Now()
+	klog.V(7).Infof("Creating http disk cache, max size: %d\n", maxSize)
+	// Walk the cache files to store initial file metadata and initial cache size.
+	fileInfos, cacheSize, err := initExistingCache(cacheDir)
+	if err != nil {
+		klog.Errorf("Error during disk cache walk: [%v]\n", err)
+	}
+	cache := &SizeLimitedDiskCache{
+		maxSize:     maxSize,
+		currentSize: cacheSize,
+		fileInfos:   fileInfos,
+		diskv: diskv.New(diskv.Options{
+			PathPerm: os.FileMode(0750),
+			FilePerm: os.FileMode(0660),
+			BasePath: cacheDir,
+			TempDir:  filepath.Join(cacheDir, ".diskv-temp"),
+		}),
+	}
+	numInitialFiles := len(fileInfos)
+	klog.V(7).Infof("Before evictions, disk cache initially contains %d files, total size: %d",
+		numInitialFiles, cacheSize)
+	beforeCacheTrim := time.Now()
+	cache.trimDiskCache(uint64(0)) // Ensure the initial disk cache size is below maxSize.
+	numFilesEvicted := numInitialFiles - len(cache.fileInfos)
+	finish := time.Now()
+	klog.V(7).Infof("%d files evicted from cache, eviction time: %s",
+		numFilesEvicted, finish.Sub(beforeCacheTrim))
+	klog.V(7).Infof("total cache initialization time: %s", finish.Sub(begin))
+	return cache
+}
+
+// initExistCache returns the slice of fileInfo for each of the files
+// in the existing cache directory, as well as the total size of
+// all these files. Sorts the fileInfo slice by ascending
+// file modification time (oldest files first). Returns an error
+// if one occurred during filepath.Walk.
+func initExistingCache(cacheDir string) ([]fileInfo, uint64, error) {
+	fileInfos := []fileInfo{}
+	currentSize := uint64(0)
+	err := filepath.Walk(cacheDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		fileSize := uint64(info.Size())
+		currentSize += fileSize
+		fi := fileInfo{
+			name:    info.Name(),
+			size:    fileSize,
+			modTime: info.ModTime(),
+		}
+		fileInfos = append(fileInfos, fi)
+		return nil
+	})
+	sort.Sort(byFileModTime(fileInfos)) // Sort by ascending modTime
+	return fileInfos, currentSize, err
+}
+
+// fileInfo stores relevant information for files in the disk cache.
+// This structure stores the size of each file to determine the
+// size of the entire disk cache, and it stores the modTime to
+// determine the oldest file to evict from the cache when necessary.
+// The name is the filename, which should be an MD5 hash (see
+// keyToFilename()).
+type fileInfo struct {
+	name    string
+	size    uint64
+	modTime time.Time
+}
+
+// Sort cache fileInfos by ascending last modification time.
+type byFileModTime []fileInfo
+
+func (fi byFileModTime) Len() int {
+	return len(fi)
+}
+
+func (fi byFileModTime) Less(i, j int) bool {
+	return fi[i].modTime.Before(fi[j].modTime)
+}
+
+func (fi byFileModTime) Swap(i, j int) {
+	fi[i], fi[j] = fi[j], fi[i]
+}
+
+// fileInfoIndex returns the index in the slice where the fileInfo
```

Comment:
Nit, you could probably have split the `fileInfoList` as its own class, especially since you have a test just for that, and the mutex isn't used outside these methods so it sort of spills into the `SizeLimitedDiskCache` class.